designs/2022-languages/README.md

```diff
@@ -0,0 +1,610 @@
+
+The [`getDirectiveComments()`](https://github.com/eslint/eslint/blob/dfc7ec11b11b56daaa10e8e6d08c5cddfc8c2c59/lib/linter/linter.js#L367) function will move into `SourceCode` to postprocess the AST with all directive comments and then expose just the disable comments through `SourceCode#getDisableDirectives()`.
+
+#### Move Rule Context Methods and Properties to `SourceCode`
+
+For JavaScript, we will need to deprecate the following methods and properties and redirect them to the `SourceCode` object:
+
+* `getAncestors()`
+* `getDeclaredVariables()`,
+* `getScope()`
+* `parserServices`
+
+The following properties we can remove once we remove the eslintrc config system, so they will still show up for JavaScript but won't be included for non-JavaScript:
+
+* `parserOptions`
+* `parserPath`
+
+#### Update Rules to New APIs
+
+With a new Rule Context, we should also start converting existing rules over to the new APIs as they move from `context` onto `SourceCode`.
+
+### Core Changes
+
+In order to make all of this work, we'll need to make the following changes in the core:
+
+1. [`linter.js`](https://github.com/eslint/eslint/blob/dfc7ec11b11b56daaa10e8e6d08c5cddfc8c2c59/lib/linter/linter.js#L1140-L1158) will need to be updated to call `SourceCode#traverse()`.
+1. `FlatConfigArray` will need to be updated to define the `language` key and to delegate validation of `languageOptions` to the language.
+1. `Linter` will need to be updated to honor the `language` key and to use JS-specific functionality where we need to provide backwards compatibility for existing JS rules. There will be a lot of code removed and delegated to the language being used, including filtering of violation via disable directives, traversing the AST, and formulating the final violation list.
+1. The `context` object passed to rules will need to be updated so it works for all languages.
+1. `RuleTester` will need to be updated to support passing in `language`. We should also updated `RuleTester` to warn about the newly-deprecated `context` methods and properties.
+
+#### Updating Rule Context
+
+The rule context object, which is passed into rules as `context`, will need to be significantly altered to support other languages. All of the methods that relate specifically to language features will need to move into `SourceCode`, leaving us with a rule context object that follows this interface:
+
+```ts
+interface RuleContext {
+
+    languageOptions: LanguageOptions;
+    settings: object;
+    options: Array<any>;
+    id: string;
+
+    getCwd(): string;
+    getFilename(): string;
+    getPhysicalFilename(): string;
+    getSourceCode(): SourceCode;
+    report(violation: Violation): void;
+}
+
+interface Violation {
+    messageId: string;
+    data: object;
+    loc: LocationRange;
+    suggest: Array<Suggestion>;
+    fix(fixer: Fixer): Iterable<FixOp>;
```

Comment:
AST-based autofixing (and suggestions) is a popular idea that's mentioned in the rewrite discussion (https://github.com/eslint/eslint/discussions/16557). It's also how various other linters do autofixing (including [ember-template-lint](https://github.com/ember-template-lint/ember-template-lint), a handlebars template linter that I help maintain. As an aside, we've been interested to use ESLint under the hood for ember-template-lint for a long time, so excited to see this RFC).

When adding support for new languages, it seems like AST-based autofixing would be a critical feature to have ready, for a number of reasons:
1. Support the migration of other linters that are already using AST-based autofixing without having to rewrite those autofixers to be based on string manipulation.
2. Avoid getting off to a bad start by requiring all the new language plugins to use string-manipulation-based autofixers that could soon be obsolete.
3. We could even decide to ONLY allow AST-based autofixing for language plugins. If that's the goal, the best time to enforce it is before any string-manipulation-based autofixers have been written. If that isn't the goal, then perhaps just make AST-based autofixing the default/recommendation for language plugins.

Reply:
The issue with ast vs string autofixing is that you need a printer to convert the AST back to code and that printer has to make decisions about where to put braces or parentheses that aren't always encoded in the AST.

For example - `const x = (1);` the AST doesn't encode the location of the `=`, the existence of the parentheses, or the existence of the semicolon (I.e `const x=  1` produces the exact same AST). Some of the info can be inferred from ranges sometimes - but a lot cannot.

So unless ESLint wants to start making stylistic decisions about code it prints after fixes - it needs a solution for that (like using a CST). Which to me is outside the scope of this particular RFC.

Reply:
A CST still might be better done before this RFC, though, for all the reasons indicated.

Reply:
There is nothing about this RFC that requires a CST, nor is there anything about this RFC that precludes AST autofixing. However, I don't think it's prudent to bundle AST mutation with this RFC or make it a dependency. What a language decides to return from `parse()` is up to it - AST, CST, whatever. Each language can decide whether or not it wants to allow AST mutation, and it may not be possible to mutate the ASTs of different languages in the same way.

Fundamentally, AST mutation just requires:

1. The ability to mutate the AST. For ESLint's architecture, this would likely need to be applied after linting vs. during traversing like in Babel. This is already the way autofixing works, so that wouldn't be an issue.
2. The ability to generate the source code given a set of AST/CST nodes.

Both of these can be added on to a language definition later.